platform/linux-generic/pktio/socket_xdp.c

```diff
@@ -163,150 +263,322 @@ static void parse_options(xdp_umem_info_t *umem_info)
+	sleep(1U);
+
+	return true;
 
-out:
-	return idx;
+err:
+	for (uint32_t j = 0U; j < i; ++j) {
+		xsk_socket__delete(sock_info->qs[j].xsk);
+		sock_info->qs[j].xsk = NULL;
+	}
+
+	return false;
 }
 
-static int sock_xdp_open(odp_pktio_t pktio, pktio_entry_t *pktio_entry, const char *devname,
-			 odp_pool_t pool_hdl)
+static void umem_delete(xdp_umem_info_t *umem_info)
 {
-	xdp_sock_info_t *priv;
-	pool_t *pool;
+	if (umem_info->ref_cnt-- != 1U)
+		return;
+
+	while (xsk_umem__delete(umem_info->umem) == -EBUSY)
+		continue;
+}
+
+static int sock_xdp_start(pktio_entry_t *pktio_entry)
+{
+	xdp_sock_info_t *priv = pkt_priv(pktio_entry);
 	int ret;
+	drv_channels_t channels = priv->q_num_conf.drv_channels;
+	struct ethtool_rxfh *indir = calloc(1U, sizeof(struct ethtool_rxfh)
+					    + sizeof(((struct ethtool_rxfh *)0)->rss_config[0U])
+					    * priv->q_num_conf.drv_num_rss);
 
-	if (disable_pktio)
+	if (indir == NULL) {
+		_ODP_ERR("Error allocating NIC RSS table\n");
 		return -1;
+	}
 
-	priv = pkt_priv(pktio_entry);
-	memset(priv, 0, sizeof(xdp_sock_info_t));
-	pool = _odp_pool_entry(pool_hdl);
-	priv->umem_info = (xdp_umem_info_t *)pool->mem_src_data;
-	ret = umem_create(priv->umem_info, pool);
+	ret = umem_create(priv->umem_info);
 
 	if (ret) {
 		_ODP_ERR("Error creating UMEM pool for xdp: %d\n", ret);
-		return -1;
+		goto err;
 	}
 
-	/* Mark transitory kernel-owned packets with the pktio index, so that they can be freed on
-	 * close. */
-	priv->pktio_idx = 1 + odp_pktio_index(pktio);
-	/* Querying with ioctl() via AF_XDP socket doesn't seem to work, so
-	 * create a helper socket for this. */
-	ret = socket(AF_INET, SOCK_DGRAM, 0);
-
-	if (ret == -1) {
-		_ODP_ERR("Error creating helper socket for xdp: %s\n", strerror(errno));
-		goto sock_err;
-	}
+	priv->q_num_conf.num_qs = _ODP_MAX(priv->q_num_conf.num_in_conf_qs,
+					   priv->q_num_conf.num_out_conf_qs);
+	priv->bind_q = priv->is_shadow_q ? priv->q_num_conf.num_qs : 0U;
+	channels.combined = priv->q_num_conf.num_qs;
 
-	priv->helper_sock = ret;
-	priv->mtu = _odp_mtu_get_fd(priv->helper_sock, devname);
+	if (!set_nic_queue_count(priv->helper_sock, pktio_entry->name, &channels))
+		_ODP_PRINT("Warning: Unable to configure NIC queue count, manual configuration"
+			   " required\n");
 
-	if (priv->mtu == 0U)
-		goto mtu_err;
+	if (priv->q_num_conf.num_in_conf_qs > 0U && !priv->is_shadow_q) {
+		indir->indir_size = priv->q_num_conf.drv_num_rss;
 
-	priv->max_mtu = pool->seg_len;
+		for (uint32_t i = 0U; i < indir->indir_size; ++i)
+			indir->rss_config[i] = (i % priv->q_num_conf.num_in_conf_qs)
+					       + priv->bind_q;
 
-	for (int i = 0; i < ODP_PKTOUT_MAX_QUEUES; ++i) {
-		odp_ticketlock_init(&priv->qs[i].rx_lock);
-		odp_ticketlock_init(&priv->qs[i].tx_lock);
+		if (!set_nic_rss_indir(priv->helper_sock, pktio_entry->name, indir))
+			_ODP_PRINT("Warning: Unable to configure NIC RSS, manual configuration"
+				   " required\n");
 	}
```

Comment:
Flow hashing still needs to be configured with external tools?

Reply:
Yes, changing these programmatically seemed quite cumbersome so NIC/driver defaults are always used. Mentioned this in `DEPENDENCIES` in v2.